### ISSUE TITLE
Problems with "?" in robots.txt

### ISSUE THREAD

**Issue by loxal** [yasserg/crawler4j#304]
In `https://www.welt.de/robots.txt` there are `?` containing entries like `Disallow: /*?config`. Hence `https://www.welt.de/test?config` should be allowed but it is not. Whereas entries like `Disallow: /*.xmli` work properly and disallow `https://www.welt.de/test.xmli`. After my investigation I figured out that `?` is the problematic character. 

I use `RobotstxtServer#allow("https://www.welt.de/test?config")` for testing.

**Comment by kkrugler** [comment]
Hi @s17t - I'm interested in why having the sitemap parsing code in the jar is an issue for you. Does it add a significant amount of code, and/or pull in a lot of other jars as transitive dependencies? Thanks!

**Comment by sebastian-nagel** [comment]
> ... `Disallow: /*?config`. Hence `https://www.welt.de/test?config` should be allowed but it is not.

According to [Google's spec](https://developers.google.com/search/reference/robots_txt) this URL is not allowed (`*` is a glob character, while `?` is not). The [RFC draft](http://www.robotstxt.org/norobots-rfc.txt) does not support any glob characters in allow/disallow statements (see [robotstxt.org](http://www.robotstxt.org/robotstxt.html)):
> Note also that globbing and regular expression are not supported ... Specifically, you cannot have lines like ... "Disallow: *.gif". 

[crawler-commons](/crawler-commons/crawler-commons) follows here the Google spec which is widely used - I suppose that even the webmaster of welt.de wants to disallow `/test?config`.

**Comment by loxal** [comment]
According to https://en.wikipedia.org/wiki/Glob_(programming) both `*` and `?` are glob characters. RFC drafts are nice but ultimately Google sets the standards (even if they are not RFC-compliant). And according to Google `*` is supported as stated in [Google's spec](https://developers.google.com/search/reference/robots_txt).

**Comment by sebastian-nagel** [comment]
Agreed, Google sets the standard and this means `?` has no glob meaning in the `Disallow:` statement which " [support a limited form of "wildcards" for path values](https://developers.google.com/search/reference/robots_txt#url-matching-based-on-path-values)".

**Comment by sebastian-nagel** [comment]
@loxal, or did I misunderstand you: Shall `/*?config` match `/test?config`or not? If yes, it's disallowed, right?

**Comment by loxal** [comment]
@sebastian-nagel whatever Google considers correct. But `?` does not need any glob meaning to support 
`Disallow: /*?config` and disallow `/test?config`? In this case `*` is a glob character and is supported, whereas `?` is just a regular character. 

I think the current implementation considers also `?` as a glob character and therefore has problems to evaluate `*?`.

**Comment by sebastian-nagel** [comment]
Ok, we agree. I haven't tested crawler4j but just wanted to make sure that crawler-commons behaves as expected: `/test?config` is forbidden but eg. `/test/config` is allowed. Thanks!

**Comment by loxal** [comment]
I just checked with https://www.google.com/webmasters/tools/robots-testing-tool
No matter what Google's documentation says, the actual implementation speaks another language:
Entries like `Disallow: /*?config` result in `/test?config` being ignored. One of those -  not too rare - case where documentation is misleading.

**Comment by sebastian-nagel** [comment]
> Entries like Disallow: /*?config result in /test?config being ignored.

What does "being ignored" mean? Is it crawled or not?

**Comment by sebastian-nagel** [comment]
If it means that Googlebot does not visit `/test?config` this is in accordance with the documentation.

**Comment by loxal** [comment]
The crawler ignores it. `/test?config` is not crawled. This is what Google's robots Tool says.

**Comment by sebastian-nagel** [comment]
Yes, but then the behavior and documentation agree: the wildcard `*` matches `test` and all remaining characters (including `?`) are matched literally. It's a disallow rule, so bots shall not crawl everything matching  `/*?config`, right?

**Comment by loxal** [comment]
Yes.

**Comment by sebastian-nagel** [comment]
I'll keep this on the radar, and will add a unit test to crawler-commons' robots.txt parser, just to make sure that it continues to work. Thanks!

**Comment by Chaiavi** [comment]
Thank you.
